File: frontend/src/api/admin_settings.ts
/**
 * api/admin_settings.ts — 系統設定相關 API（僅管理員）
 *
 * 管理 SMTP 郵件設定、App URL、通知排程時間等全域設定。
 * 設定儲存在後端資料庫的 system_settings 表，優先權高於 .env 設定。
 *
 * 注意：smtp_password 欄位特殊處理：
 *  - GET 回傳 smtp_password_set（boolean），不回傳實際密碼
 *  - PUT 傳空字串或不傳 = 保留現有密碼；傳有值才更新
 *  - encryption_key_configured 為 false 時表示後端未設定加密金鑰，密碼無法透過 UI 修改
 */
import type { AxiosError } from 'axios'
import { api } from './client'
import type { ApiResponse } from '@/types/api'

/** 後端回傳的系統設定結構，smtp_password 只有已設定/未設定狀態，不回傳明文 */
export interface SystemSettings {
  smtp_host: string
  smtp_port: number
  smtp_user: string
  smtp_password_set: boolean
  smtp_from: string
  smtp_sender_name: string
  app_url: string
  notification_cron_hour: number
  notification_cron_minute: number
  encryption_key_configured: boolean
}

/** 更新設定的 payload，所有欄位皆為選填，只傳需要更新的欄位 */
export interface SettingsUpdatePayload {
  smtp_host?: string
  smtp_port?: number
  smtp_user?: string
  smtp_password?: string
  smtp_from?: string
  smtp_sender_name?: string
  app_url?: string
  notification_cron_hour?: number
  notification_cron_minute?: number
}

/** 測試寄信用的 SMTP 設定，使用表單目前填入的值（不需要先儲存） */
export interface TestEmailPayload {
  smtp_host: string
  smtp_port: number
  smtp_user: string
  smtp_password?: string
  smtp_from: string
  smtp_sender_name: string
}

/**
 * 從 Axios 錯誤回應中提取後端的錯誤訊息。
 * 後端可能回傳 detail（FastAPI 格式）或 message（ApiResponse 格式），優先取 detail。
 */
function extractMessage(err: unknown, fallback: string): never {
  const detail = (err as AxiosError<{ detail?: string }>)?.response?.data?.detail
  const message = detail ?? (err as AxiosError<ApiResponse<null>>)?.response?.data?.message ?? fallback
  throw new Error(message)
}

/** 取得目前的系統設定 */
export async function getSystemSettings(): Promise<SystemSettings> {
  const { data } = await api.get<ApiResponse<SystemSettings>>('/api/v1/admin/settings')
  if (!data.success || !data.data) throw new Error(data.message)
  return data.data
}

/**
 * 儲存系統設定。
 * smtp_password 傳 undefined 表示不變更密碼；傳空字串也視為不變更。
 */
export async function updateSystemSettings(payload: SettingsUpdatePayload): Promise<void> {
  try {
    const { data } = await api.put<ApiResponse<null>>('/api/v1/admin/settings', payload)
    if (!data.success) throw new Error(data.message)
  } catch (err) {
    if ((err as AxiosError)?.response) {
      return extractMessage(err, '儲存設定失敗')
    }
    throw err instanceof Error ? err : new Error('網路錯誤，請確認連線後重試')
  }
}

/**
 * 使用表單填入的 SMTP 設定發送測試信到 smtp_from 信箱。
 * 不需要先儲存設定，方便在儲存前先確認 SMTP 是否正常。
 * 成功時後端回傳成功訊息字串。
 */
export async function testSmtpEmail(payload: TestEmailPayload): Promise<string> {
  try {
    const { data } = await api.post<ApiResponse<null>>('/api/v1/admin/settings/test-email', payload)
    if (!data.success) throw new Error(data.message)
    return data.message
  } catch (err) {
    if ((err as AxiosError)?.response) {
      return extractMessage(err, '測試寄信失敗')
    }
    throw err instanceof Error ? err : new Error('網路錯誤，請確認連線後重試')
  }
}

File: backend/src/api/v1/routers/admin_settings.py
import logging
import smtplib

from application.services.settings_service import SettingsService
from domain.entities.user import User
from domain.exceptions import BadRequestException
from fastapi import APIRouter, Depends
from infrastructure.smtp.smtp_email_sender import SmtpEmailSender

from api.dependencies import get_settings_service, require_admin
from api.v1.schemas.admin_settings import (
    SettingsResponse,
    SettingsUpdateRequest,
    TestEmailRequest,
)
from api.v1.schemas.base import ApiResponse

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v1/admin/settings", tags=["admin-settings"])


@router.get("", response_model=ApiResponse[SettingsResponse])
async def get_settings(
    _: User = Depends(require_admin),
    svc: SettingsService = Depends(get_settings_service),
) -> ApiResponse[SettingsResponse]:
    s = await svc.get_all_settings()
    return ApiResponse.ok(
        data=SettingsResponse(
            smtp_host=s.get("smtp_host") or "",
            smtp_port=int(s.get("smtp_port") or "587"),
            smtp_user=s.get("smtp_user") or "",
            smtp_password_set=bool(s.get("smtp_password")),
            smtp_from=s.get("smtp_from") or "",
            smtp_sender_name=s.get("smtp_sender_name") or "SubTrack",
            app_url=s.get("app_url") or "",
            notification_cron_hour=int(s.get("notification_cron_hour") or "8"),
            notification_cron_minute=int(s.get("notification_cron_minute") or "0"),
            encryption_key_configured=svc.encryption_key_configured,
        )
    )


@router.put("", response_model=ApiResponse[None])
async def update_settings(
    body: SettingsUpdateRequest,
    _: User = Depends(require_admin),
    svc: SettingsService = Depends(get_settings_service),
) -> ApiResponse[None]:
    if body.smtp_password:
        if not svc.encryption_key_configured:
            raise BadRequestException("加密金鑰未設定（SETTINGS_ENCRYPTION_KEY），無法儲存密碼")
        await svc.set("smtp_password", body.smtp_password)

    field_map: dict[str, str | None] = {
        "smtp_host": body.smtp_host,
        "smtp_port": str(body.smtp_port) if body.smtp_port is not None else None,
        "smtp_user": body.smtp_user,
        "smtp_from": body.smtp_from,
        "smtp_sender_name": body.smtp_sender_name,
        "app_url": body.app_url,
        "notification_cron_hour": str(body.notification_cron_hour)
        if body.notification_cron_hour is not None
        else None,
        "notification_cron_minute": str(body.notification_cron_minute)
        if body.notification_cron_minute is not None
        else None,
    }
    for key, value in field_map.items():
        if value is not None:
            await svc.set(key, value)

    return ApiResponse.ok(message="設定已儲存")


@router.post("/test-email", response_model=ApiResponse[None])
async def test_email(
    body: TestEmailRequest,
    current_user: User = Depends(require_admin),
    svc: SettingsService = Depends(get_settings_service),
) -> ApiResponse[None]:
    password = body.smtp_password
    if not password:
        password = await svc.get("smtp_password") or ""

    sender = SmtpEmailSender(
        host=body.smtp_host,
        port=body.smtp_port,
        username=body.smtp_user,
        password=password,
        from_addr=body.smtp_from,
        sender_name=body.smtp_sender_name,
    )
    try:
        await sender.send(
            to=[current_user.email],
            subject="SubTrack 郵件設定測試",
            body=(
                "這是一封測試信，確認您的 SMTP 設定正確。\n\n"
                "如果您收到這封信，表示 SMTP 設定無誤。"
            ),
        )
    except (smtplib.SMTPException, OSError) as e:
        logger.warning("SMTP test failed: %s", e)
        raise BadRequestException("寄信失敗，請確認 SMTP 設定是否正確")

    return ApiResponse.ok(message=f"測試信已寄至 {current_user.email}")
